staging_service/import_specifications/file_writers.py

```diff
@@ -0,0 +1,188 @@
+                raise ImportSpecWriteException(
+                    err + f" does not have the same keys as {_ORDER_AND_DISPLAY}")
+            for pid, v in datarow.items():
+                if v is not None and not isinstance(v, numbers.Number) and not isinstance(v, str):
+                    raise ImportSpecWriteException(
+                        err + f"'s value for parameter {pid} is not a number or a string")
+
+
+def _check_string(tocheck: Any, errprefix: str):
+    if not isinstance(tocheck, str) or not tocheck.strip():
+        raise ImportSpecWriteException(
+            errprefix + " cannot be a non-string or a whitespace only string")
+
+
+def _check_is_sequence(tocheck: Any, errprefix: str):
+    if not (isinstance(tocheck, collections.abc.Sequence) and not isinstance(tocheck, str)):
+        raise ImportSpecWriteException(errprefix + " is not a list")
+
+
+# TODO WRITE_XSV look into server OOM protection if the user sends a huge JSON packet
+
+def write_csv(folder: Path, types: dict[str, dict[str, list[Any]]]) -> dict[str, Path]:
+    """
+    Writes import specifications to 1 or more csv files. All the writers in this module
+    have the same function signatures; see the module level documentation.
+    """
+    return _write_xsv(folder, types, _EXT_CSV, _SEP_CSV)
+
+
+def write_tsv(folder: Path, types: dict[str, dict[str, list[Any]]]) -> dict[str, Path]:
+    """
+    Writes import specifications to 1 or more tsv files. All the writers in this module
+    have the same function signatures; see the module level documentation.
+    """
+    return _write_xsv(folder, types, _EXT_TSV, _SEP_TSV)
+
+
+def _write_xsv(folder: Path, types: dict[str, dict[str, list[Any]]], ext: str, sep: str):
+    _check_write_args(folder, types)
+    res = {}
+    for datatype in types:
+        out = folder / (datatype + "." + ext)
+        dt = types[datatype]
+        cols = len(dt[_ORDER_AND_DISPLAY])
+        with open(out, "w", newline='') as f:
+            csvw = csv.writer(f, delimiter=sep)  # handle sep escaping
+            csvw.writerow([f"{_DATA_TYPE} {datatype}{_HEADER_SEP} "
+                           + f"{_COLUMN_STR} {cols}{_HEADER_SEP} {_VERSION_STR} {_VERSION}"])
+            pids = [i[0] for i in dt[_ORDER_AND_DISPLAY]]
+            csvw.writerow(pids)
+            csvw.writerow([i[1] for i in dt[_ORDER_AND_DISPLAY]])
+            for row in dt[_DATA]:
+                csvw.writerow([row[pid] for pid in pids])
+        res[datatype] = out
+    return res
+
+
+def _check_write_args(folder: Path, types: dict[str, dict[str, list[Any]]]):
+    if not folder:
+        # this is a programming error, not a user input error, so not using the custom
+        # exception here
+        raise ValueError("The folder cannot be null")
+    if type(types) != dict:
+        raise ImportSpecWriteException("The types value must be a mapping")
+    _check_import_specification(types)
+
+
+class ImportSpecWriteException(Exception):
```

Comment:
Probably YAGNI, but we may want to consider creating an StagingServiceException(Exception) and then deriving new exceptions from StagingServiceException instead of Exception, to allow for catching Staging Service Exceptions and Exceptions 
```
try:

except StagingServiceException

except Execption
```

Reply:
Sounds good, made an issue. Next time I touch the code base I'll do it